File: BasicSample/app/src/main/java/com/example/android/persistence/DataRepository.java
package com.example.android.persistence;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import com.example.android.persistence.db.AppDatabase;
import com.example.android.persistence.db.entity.CommentEntity;
import com.example.android.persistence.db.entity.ProductEntity;

import java.util.List;

/**
 * Repository handling the work with products and comments.
 */
public class DataRepository {

    private static DataRepository sInstance;
    //持有的数据库实例
    private final AppDatabase mDatabase;
    private MediatorLiveData<List<ProductEntity>> mObservableProducts;

    private DataRepository(final AppDatabase database) {
        mDatabase = database;
        mObservableProducts = new MediatorLiveData<>();
        //                              room 加载全部产品数据
        mObservableProducts.addSource(mDatabase.productDao().loadAllProducts(),
                productEntities -> {
                    //产品不为空
                    if (mDatabase.getDatabaseCreated().getValue() != null) {
                        mObservableProducts.postValue(productEntities);
                    }
                });
    }
    //获取 Repository 实例
    public static DataRepository getInstance(final AppDatabase database) {
        if (sInstance == null) {
            synchronized (DataRepository.class) {
                if (sInstance == null) {
                    sInstance = new DataRepository(database);
                }
            }
        }
        return sInstance;
    }

    /**
     * Get the list of products from the database and get notified when the data changes.
     */
    public LiveData<List<ProductEntity>> getProducts() {
        return mObservableProducts;
    }

    public LiveData<ProductEntity> loadProduct(final int productId) {
        return mDatabase.productDao().loadProduct(productId);
    }

    public LiveData<List<CommentEntity>> loadComments(final int productId) {
        return mDatabase.commentDao().loadComments(productId);
    }
    //搜索
    public LiveData<List<ProductEntity>> searchProducts(String query) {
        return mDatabase.productDao().searchAllProducts(query);
    }
}

File: BasicSample-kotlin/app/src/main/java/com/kotlin/basicsample_kotlin/DataRepository.kt
package com.kotlin.basicsample_kotlin

import androidx.lifecycle.MediatorLiveData
import com.kotlin.basicsample_kotlin.db.AppDatabase
import com.kotlin.basicsample_kotlin.db.entity.ProductEntity

class DataRepository(var mdatabase: AppDatabase) {
    private var mObservableProducts: MediatorLiveData<List<ProductEntity>> = MediatorLiveData()

    init {
        mObservableProducts.addSource(mdatabase.productDao().loadAllProducts()) { productEntities ->
            mdatabase.getDatabaseCreated()?.let {
                mObservableProducts.postValue(productEntities)
            }
        }
    }

    companion object {
        private var sInstance: DataRepository? = null
        fun getInstance(database: AppDatabase): DataRepository {
            sInstance?.let { return sInstance as DataRepository }
            return DataRepository(database)
        }
    }

    fun getProducts() = mObservableProducts
    fun loadProduct(productId: Int) = mdatabase.productDao().loadProduct(productId)
    fun loadComments(productId: Int) = mdatabase.commentDao().loadComments(productId)
    fun searchProducts(query: String) = mdatabase.productDao().searchAllProducts(query)
}